addons/website_sale/static/src/js/website_sale_recently_viewed.js

```diff
@@ -0,0 +1,181 @@
+                    products: products,
+                    uniqueId: this.uniqueId,
+                    productFrame: 4,
+                    currency: currency,
+                    wishlist: wishlist,
+                    compare: compare,
+                    rating: rating,
+                    widget: this,
+                }));
+                this.mobileCarousel = $(qweb.render('website_sale.productsRecentlyViewed', {
+                    products: products,
+                    uniqueId: this.uniqueId,
+                    productFrame: 1,
+                    currency: currency,
+                    wishlist: wishlist,
+                    compare: compare,
+                    rating: rating,
+                    widget: this,
+                }));
+                this._addCarousel();
+            }
+        }
+    },
+
+    /**
+     * @private
+     */
+    _addCarousel: function () {
+        this.$el.find('.slider').html(config.device.size_class <= config.device.SIZES.SM ? this.mobileCarousel : this.webCarousel);
+        // Only way to stop this bs carousel...
+        this.$el.find('#' + this.uniqueId).carousel('pause');
+    },
+
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    _onAddToCart: function (ev) {
+        var self = this;
+        var $card = $(ev.currentTarget).closest('.card');
+        this._rpc({
+            route: "/shop/cart/update_json",
+            params: {
+                product_id: $card.find('input[data-product-id]').data('product-id'),
+                add_qty: 1
+            },
+        }).then(function (data) {
+            var $q = $(".my_cart_quantity");
+            if (data.cart_quantity) {
+                $q.parents('li:first').removeClass('d-none');
+                $q.html(data.cart_quantity).hide().fadeIn(600);
+            }
+
+            var $navButton = wSaleUtils.getNavBarButton('.o_wsale_my_cart');
+            wSaleUtils.animateClone($navButton, $(ev.currentTarget).parents('.o-product-card'), 25, 40);
+
+            self._dp.add(self._fetch()).then(self._render.bind(self));
+
+            if ($('.oe_cart')) {
+                $(".js_cart_lines").first().before(data['website_sale.cart_lines']).end().remove();
+                $(".js_cart_summary").first().before(data['website_sale.short_cart_summary']).end().remove();
+            }
+        });
+    },
+});
+
+var productsRecentlyViewedUpdate = publicWidget.Widget.extend({
+    selector: '#product_details',
+    events: {
+        'change input.product_id[name="product_id"]': '_onProductChange',
+    },
+    debounceValue: 3000,
+
+    /**
+     * @constructor
+     */
+    init: function () {
+        this._super.apply(this, arguments);
+        this._onProductChange = _.debounce(this._onProductChange, this.debounceValue);
```

Comment:
and it will avoid all duplicate record to aggregate later

when do you plan to aggregate data ? 
![image](https://user-images.githubusercontent.com/4848797/63413606-56fa5f00-c3fa-11e9-8fd6-504ff3f9407a.png)

Idem for homepage? do we plan to keep 1 pageview, by homepage visit? (@dbeguin )
or maybe something to improvein " # Add page even if already in website_track_ids as checks on relations are done in many2many write method"

Reply:
What I did in my implementation was to put a constraint on user/product to be unique and update it, if it fails I create the record  and except IntegrityError in case. This way there is no duplicate, I don't know if there is a more "odoo" way of doing it.

```
with self.pool.cursor() as pv_cr:
            pv_cr.execute('''UPDATE visitor SET blablabla'')
            fetch = pv_cr.fetchone()
            if fetch:
                return True
            else:
                # update failed
                try:
                    with tools.mute_logger('odoo.sql_db'):
                        pv_cr.execute('''
                            INSERT INTO visitor blablabla
                            RETURNING id;
                            ''')
                    fetch = pv_cr.fetchone()
                    if fetch:
                        return True
                except IntegrityError:
                    return False```

Reply:
as you cann see in screenshot it is not the case for now...

but I don't check more, but whats is strange, is that comment let think that it is handled: # Add page even if already in website_track_ids as checks on relations are done in many2many write method